lsp-completion.el

```diff
@@ -0,0 +1,542 @@
+            (-let [(callback cleanup-fn) (lsp--create-apply-text-edits-handlers)]
+              (lsp--resolve-completion-async
+               item
+               (lambda (resolved-item)
+                 (funcall callback
+                          (lsp:completion-item-additional-text-edits? resolved-item)))
+               cleanup-fn))))
+
+        (when (and lsp-signature-auto-activate
+                   (lsp-feature? "textDocument/signatureHelp"))
+          (lsp-signature-activate))
+
+        (setq-local lsp-inhibit-lsp-hooks nil)
+
+        (when (lsp-completion--looking-back-trigger-characterp trigger-chars)
+          (setq this-command 'self-insert-command)))
+    (lsp-completion--capf-clear-cache)))
+
+(defun lsp-completion--setup-company ()
+  "Setup company-mode."
+  (cond
+   ((and (functionp 'company-lsp)
+         (not lsp-completion-prefer-capf))
+    (progn
+      (company-mode 1)
+      (add-to-list 'company-backends 'company-lsp)
+      (setq-local company-backends (remove 'company-capf company-backends))))
+
+   ((and (fboundp 'company-mode) lsp-completion-enable)
+    (company-mode 1)
+    (add-to-list 'company-backends 'company-capf)))
+
+  (add-hook 'company-completion-started-hook
+            (lambda (&rest _)
+              (setq-local lsp-inhibit-lsp-hooks t))
+            nil
+            t)
+  (add-hook 'company-after-completion-hook
+            (lambda (&rest _)
+              (lsp-completion--capf-clear-cache)
+              (setq-local lsp-inhibit-lsp-hooks nil))
+            nil
+            t))
+
+(defun lsp-completion--clean-company ()
+  "Clean company-mode."
+  (remove-hook 'company-completion-started-hook
+               (lambda (&rest _)
+                 (setq-local lsp-inhibit-lsp-hooks t))
+               t)
+  (remove-hook 'company-after-completion-hook
+               (lambda (&rest _)
+                 (lsp-completion--capf-clear-cache)
+                 (setq-local lsp-inhibit-lsp-hooks nil))
+               t))
+
+(defun lsp-completion--enable ()
+  "Enable LSP completion support."
+  (when (and lsp-completion-enable
+             (lsp-feature? "textDocument/completion"))
+    (setq-local completion-at-point-functions nil)
+    (add-hook 'completion-at-point-functions #'lsp-completion-at-point nil t)
+    (setq-local completion-category-defaults
+                (add-to-list 'completion-category-defaults '(lsp-capf (styles basic))))
+    (lsp-completion-mode 1)))
+
+(defun lsp-completion--disable ()
+  "Disable LSP completion support."
+  (lsp-completion-mode -1)
+  (remove-hook 'completion-at-point-functions #'lsp-completion-at-point t)
+  (setq-local completion-category-defaults
+              (cl-remove 'lsp-capf completion-category-defaults :key #'car)))
+
+;;;###autoload
+(define-minor-mode lsp-completion-mode
+  "Toggle LSP completion support."
+  :group 'lsp-mode
+  :global nil
+  :lighter ""
+  (cond
+   (lsp-completion-mode
+    (when (bound-and-true-p company-mode)
```

Comment:
This one is not correct, since `company-mode` and completion are loosely coupled.
Before even without `company-mode` user can still get completion via `completion-at-point`, now since this check, the completion at point functions will not be added

Reply:
Good to know, I thought company-mode was required to use completion.
I moved out the 2 hooks of this check.